File: AtCoder/GrandContest12/B/b.py
#! /usr/bin/env python3
from scipy.sparse import lil_matrix, csc_matrix, identity
from scipy.sparse.linalg import inv
import numpy as np
import sys

def query(mat, v, d, c, color):
    # acc1 = acc2 = identity(n, format='csc')
    # for i in range(d):
    #     acc1 = acc1 * mat
    #     acc2 = acc2 + acc1

    E = identity(n, format='csc')
    acc2 = inv(E - mat) * (E - mat**(d+1))
    
    vec = acc2.getcol(v-1)
    mask = vec.sign()
    color = color - color.multiply(mask)
    color = color + c*mask
    return color

data = map(int, sys.stdin.read().split())

n = next(data)
m = next(data)
mat = lil_matrix((n, n))

for i in range(m):
    a = next(data)
    b = next(data)
    mat[a-1, b-1] = 1
    mat[b-1, a-1] = 1
mat = mat.tocsc()

q = next(data)
color = csc_matrix([[0] for i in range(n)])

for i in range(q):
    v = next(data)
    d = next(data)
    c = next(data)
    color = query(mat, v, d, c, color)

for i in range(n):
    print(int(color[i,0]))

File: AtCoder/GrandContest12/B/b.cpp
#include <stdio.h>
#include <vector>

using namespace std;

typedef struct _Node {
    int color = 0;
    vector<struct _Node*> next;
    int d = -1;
} Node;

void query(Node* node, int d, int c) {
    if (node->d >= d) {
        return ;
    }
    
    node->d = d;

    if (node->color == 0) {
        node->color = c;
    }
    
    for (vector<Node*>::iterator it = node->next.begin();
         it != node->next.end(); it++) {
        query((*it), d-1, c);
    }
}

int main(void) {
    int n, m, q;
    int *a, *b, *v, *d, *c;
    Node *nodes;

    scanf("%d %d", &n, &m);
    
    nodes = new Node[n];
    a = new int[m];
    b = new int[m];

    for (int i = 0; i < m; i++) {
        scanf("%d %d", &a[i], &b[i]);
        nodes[a[i]-1].next.push_back(&nodes[b[i]-1]);
        nodes[b[i]-1].next.push_back(&nodes[a[i]-1]);
    }

    scanf("%d", &q);

    v = new int[q];
    d = new int[q];
    c = new int[q];

    for (int i = 0; i < q; i++) {
        scanf("%d %d %d", &v[i], &d[i], &c[i]);
    }

    for (int i = q-1; i >= 0; i--) {
        query(&nodes[v[i]-1], d[i], c[i]);
    }

    for (int i = 0; i < n; i++) {
        printf("%d\n", nodes[i].color);
    }

    return 0;
}
